pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -54,166 +57,201 @@ func NewDockercfgController(cl client.Interface, options DockercfgControllerOpti
-		return nil
-
-	case foundPullSecret && !foundMountableSecret, !foundPullSecret && foundMountableSecret:
-		dockercfgSecretName := ""
+	if foundPullSecret || foundMountableSecret {
 		switch {
 		case foundPullSecret:
-			dockercfgSecretName = imageDockercfgPullSecrets.List()[0]
+			serviceAccount.Secrets = append(serviceAccount.Secrets, api.ObjectReference{Name: imageDockercfgPullSecrets.List()[0]})
 		case foundMountableSecret:
-			dockercfgSecretName = mountableDockercfgSecrets.List()[0]
+			serviceAccount.ImagePullSecrets = append(serviceAccount.ImagePullSecrets, api.LocalObjectReference{Name: mountableDockercfgSecrets.List()[0]})
 		}
 
-		err := e.createDockerPullSecretReference(serviceAccount, dockercfgSecretName)
-		if kapierrors.IsConflict(err) {
-			// nothing to do.  Our choice was stale or we got a conflict.  Either way that means that the service account was updated.  We simply need to return because we'll get an update notification later
-			return nil
-		}
-
-		return err
-
-	}
-
-	// if we get here, then we need to create a new dockercfg secret
-	// before we do expensive things, make sure our view of the service account is up to date
-	if liveServiceAccount, err := e.client.ServiceAccounts(serviceAccount.Namespace).Get(serviceAccount.Name); err != nil {
+		_, err = e.client.ServiceAccounts(serviceAccount.Namespace).Update(serviceAccount)
 		return err
-	} else if liveServiceAccount.ResourceVersion != serviceAccount.ResourceVersion {
-		// our view of the service account is not up to date
-		// we'll get notified of an update event later and get to try again
-		// this only prevent interactions between successive runs of this controller's event handlers, but that is useful
-		glog.V(2).Infof("View of ServiceAccount %s/%s is not up to date, skipping dockercfg creation", serviceAccount.Namespace, serviceAccount.Name)
-		return nil
 	}
 
 	dockercfgSecret, err := e.createDockerPullSecret(serviceAccount)
 	if err != nil {
 		return err
 	}
 
-	err = e.createDockerPullSecretReference(serviceAccount, dockercfgSecret.Name)
-	if kapierrors.IsConflict(err) {
-		// nothing to do.  Our choice was stale or we got a conflict.  Either way that means that the service account was updated.  We simply need to return because we'll get an update notification later
-		// we do need to clean up our dockercfgSecret.  token secrets are cleaned up by the controller handling service account dockercfg secret deletes
-		glog.V(2).Infof("Deleting secret %s/%s (err=%v)", dockercfgSecret.Namespace, dockercfgSecret.Name, err)
-		if err := e.client.Secrets(dockercfgSecret.Namespace).Delete(dockercfgSecret.Name); (err != nil) && !kapierrors.IsNotFound(err) {
-			utilruntime.HandleError(err)
+	first := true
+	err = client.RetryOnConflict(client.DefaultBackoff, func() error {
+		if !first {
+			obj, exists, err := e.serviceAccountCache.GetByKey(key)
+			if err != nil {
+				return err
+			}
+			if !exists || !needsDockercfgSecret(serviceAccount) {
```

Comment:
this is checking the `serviceAccount` we processed last time through the loop, which will always have the dockercfg secret added to it (but not actually persisted). the `!needsDockercfgSecret` check should be done on the retrieved `obj`